### PR TITLE
Fix #542: Return more complete detail of inbox message

### LINKED ISSUE #542: Return more complete detail of inbox message

### PR DESCRIPTION
**Pull request by petrdvorak**

### CODE REVIEW

File: powerauth-push-model/src/main/java/io/getlime/push/model/request/CreateInboxMessageRequest.java [hunk 1]
```diff
@@ -42,4 +47,8 @@ public class CreateInboxMessageRequest {
     @ToString.Exclude
     private String body;
 
+    @NotNull
+    @NotEmpty
+    private List<@NonNull @Pattern(regexp="^[a-zA-Z0-9_-]{3,255}$") String> applications = new ArrayList<>();
```
Comment (romanstrobl) [inline powerauth-push-model/src/main/java/io/getlime/push/model/request/CreateInboxMessageRequest.java:52]: Are you sure you want to use `lombok.NonNull` annotation in the list?

File: powerauth-push-server/src/main/java/io/getlime/push/repository/InboxRepository.java [hunk 2]
```diff
@@ -37,47 +38,47 @@ public interface InboxRepository extends PagingAndSortingRepository<InboxMessage
...
+    long countAllByUserIdAndApplicationsContainingAndRead(String userId, AppCredentialsEntity app, boolean read);
 
     /**
      * Mark all user messages as read.
      * @param userId User ID.
-     * @param appId Application ID.
+     * @param app Application.
      */
-    @Query("UPDATE InboxMessageEntity i SET i.read = true WHERE i.userId = :userId AND i.appId = :appId AND i.read = false")
+    @Query("UPDATE InboxMessageEntity i SET i.read = true WHERE i.userId = :userId AND :app member i.applications AND i.read = false")
```
Comment (romanstrobl) [inline powerauth-push-server/src/main/java/io/getlime/push/repository/InboxRepository.java:80]: JPAQL provides the `MEMBER OF` operator, does `MEMBER` operator really work?

**Review by romanstrobl — CHANGES_REQUESTED** [review]
I found two issues in the code which should be resolved.

File: docs/sql/mysql/create_push_server_schema.sql [hunk 3]
```diff
@@ -87,5 +92,7 @@ CREATE INDEX `push_campaign_user_campaign` ON `push_campaign_user`(`campaign_id`
 CREATE INDEX `push_campaign_user_detail` ON `push_campaign_user`(`user_id`);
 
 CREATE INDEX `push_inbox_id` ON `push_inbox` (`inbox_id`);
-CREATE INDEX `push_inbox_user_app` ON `push_inbox` (`user_id`, `app_id`);
+CREATE INDEX `push_inbox_user_app` ON `push_inbox` (`user_id`);
```
Comment (romanstrobl) [inline docs/sql/mysql/create_push_server_schema.sql:95]: There is not `app` in the `push_inbox_user_app` index, is the index name correct?

File: docs/sql/mysql/create_push_server_schema.sql [hunk 4]
```diff
@@ -87,5 +92,7 @@ CREATE INDEX `push_campaign_user_campaign` ON `push_campaign_user`(`campaign_id`
...
 CREATE INDEX `push_inbox_user_read` ON `push_inbox` (`user_id`, `read`);
+
+CREATE INDEX `push_inbox_app_inbox` ON `push_inbox_app` (`inbox_id`, `app_credentials_id`);
```
Comment (romanstrobl) [inline docs/sql/mysql/create_push_server_schema.sql:98]: This index seems a bit strange, it should be unique, we probably do not want to allow duplicates. Probably a good candidate for a primary key.

File: docs/sql/mysql/create_push_server_schema.sql [hunk 5]
```diff
@@ -68,6 +67,12 @@ CREATE TABLE `push_inbox` (
   PRIMARY KEY (`id`)
 ) ENGINE=InnoDB AUTO_INCREMENT=1 CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;
 
+-- Create table for assignment of inbox messages to apps
+CREATE TABLE `push_inbox_app` (
```
Comment (romanstrobl) [inline docs/sql/mysql/create_push_server_schema.sql:71]: The table has no primary key. Usually database tables do have a primary key, unless it is just a dump of data.

**Review by romanstrobl — CHANGES_REQUESTED** [review]
I also noticed that the new DB table `push_inbox_app` is missing a primary key, which is a bit of an antipattern in database design for structured data.

File: powerauth-push-model/src/main/java/io/getlime/push/model/request/CreateInboxMessageRequest.java (hunk 1 above)
Comment (petrdvorak) [reply powerauth-push-model/src/main/java/io/getlime/push/model/request/CreateInboxMessageRequest.java:52]: good catch

File: powerauth-push-server/src/main/java/io/getlime/push/repository/InboxRepository.java (hunk 2 above)
Comment (petrdvorak) [reply powerauth-push-server/src/main/java/io/getlime/push/repository/InboxRepository.java:80]: Strangely, this seems to work. I will change it to `MEMBER OF`.

File: docs/sql/mysql/create_push_server_schema.sql (hunk 3 above)
Comment (petrdvorak) [reply docs/sql/mysql/create_push_server_schema.sql:95]: Thank you for catching this, issue for MySQL and Oracle. PostgreSQL was correct.

File: docs/sql/mysql/create_push_server_schema.sql (hunk 4 above)
Comment (petrdvorak) [reply docs/sql/mysql/create_push_server_schema.sql:98]: Absolutely, I will fix this.

File: docs/sql/mysql/create_push_server_schema.sql (hunk 5 above)
Comment (petrdvorak) [reply docs/sql/mysql/create_push_server_schema.sql:71]: Agreed.

**Review by romanstrobl — APPROVED** [review]
Fix approved.

**Pull request merged**